### PR TITLE
Update NVDA Controller client to support speaking SSML sequences

### LINKED ISSUE #5638: Add the ability to determine if NVDA is speaking in NVDA Controller.dll
See title.

### LINKED ISSUE #11028: NVDA controller client: allow choosing speech prioiryty when speaking
### Is your feature request related to a problem? Please describe.
External software can use the NVDA controller client library to make NVDA announce text. This is used by twitter clients like TWBlue and Chicken Nugget, for example. This functionality is also used internally, for example by the live region handler.

However, the new speech priority support is not added to the NVDA controller client. It would be helpful if it were,  See #10622 for a use case.

### Describe the solution you'd like
I propose adding a nvdaController_speakTextEx function that adds:

* An enum parameter for the speech priority
* A boolean value that allows one to cancel speech before speaking. I think it should only cancel speech within the speech queue of the given priority. If someone wants to cancel all the speech, they can use nvdaController_cancelSpeech instead.

### Describe alternatives you've considered
Add these new parameters to nvdaController_speakText. I'm afraid this can't be done without breaking backwards compatibility though. Could you confirm this @michaelDCurran?

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Fixes #11028
Fixes #5638

### Summary of the issue:
The NVDA Controller client has been stable for a long time, but it lacked support for modern speech features, such as priority and callbacks.

### Description of user facing changes
None.

### Description of development approach
Added the following functions to the controller client:
- `nvdaController_getProcessId`: To get the process id (PID) of the current instance of NVDA the controller client is using.
- `nvdaController_speakSsml`: To instruct NVDA to speak according to the given SSML. This function also supports:
    - Providing the symbol level.
    - Providing the priority of speech to be spoken.
    - speaking both synchronously (blockking) and asynchronously (instant return).
- `nvdaController_setOnSsmlMarkReachedCallback`: To register a callback of type `onSsmlMarkReachedFuncType` that is called in synchronous mode for every `<mark />` tag encountered in the SSML sequence provided to `nvdaController_speakSsml`.

### Testing strategy:
Updated python example. Note that this doesn't work very well with Espeak, but that's more of espeaks fault. This is best tested with OneCore.

### Known issues with pull request:
- NVDA Helper Remote exports functions to behave like the NVDA Controller client. It looks like these exports are used nowhere in the wild. I kept this for now, but am tempted to remove it from NVDAHelperRemote.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

### CODE REVIEW

File: source/NVDAHelper.py [hunk 1]
```diff
@@ -76,7 +155,8 @@ def nvdaController_cancelSpeech():
 	import queueHandler
 	import speech
 	queueHandler.queueFunction(queueHandler.eventQueue,speech.cancelSpeech)
-	return 0
+	return SystemErrorCodes.SUCCESS
+
 
 @WINFUNCTYPE(c_long,c_wchar_p)
 def nvdaController_brailleMessage(text):
```
Comment (seanbudd) [inline source/NVDAHelper.py:162]:
```suggestion
def nvdaController_brailleMessage(text: str) -> SystemErrorCodes:
```

File: source/NVDAHelper.py [hunk 2]
```diff
@@ -63,10 +73,79 @@ def nvdaController_speakText(text):
...
+	return SystemErrorCodes.SUCCESS
+
+
+@WINFUNCTYPE(c_long, c_wchar_p, c_int, c_int, c_bool)
+def nvdaController_speakSsml(
+		ssml: str,
+		symbolLevel: int,
+		priority: "SpeechPriority",
+		asynchronous: bool,
+):
```
Comment (seanbudd) [inline source/NVDAHelper.py:87]:
```suggestion
) -> int | SystemErrorCodes:
```

File: source/NVDAHelper.py [hunk 3]
```diff
@@ -76,7 +155,8 @@ def nvdaController_cancelSpeech():
 	import queueHandler
```
Comment (seanbudd) [inline source/NVDAHelper.py:155]: should queueHandler be removed here too?

File: nvdaHelper/interfaces/nvdaController/nvdaController.idl [hunk 4]
```diff
@@ -26,12 +26,39 @@ cpp_quote("This license can be found at:")
 cpp_quote("http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html")
 cpp_quote("*/")
 
+typedef enum {
```
Comment (seanbudd) [inline nvdaHelper/interfaces/nvdaController/nvdaController.idl:29]: could you add a comment on this Enum that it must match the enum in python code.
Same with in the python definition, so that these 2 enums are kept in sync.

File: nvdaHelper/interfaces/nvdaController/nvdaController.idl [hunk 5]
```diff
@@ -26,12 +26,39 @@ cpp_quote("This license can be found at:")
...
+} SPEECH_PRIORITY;
+
+typedef enum {
+	SYMBOL_LEVEL_NONE = 0,
+	SYMBOL_LEVEL_SOME = 100,
+	SYMBOL_LEVEL_MOST = 200,
+	SYMBOL_LEVEL_ALL = 300,
+	SYMBOL_LEVEL_CHAR = 1000,
+	SYMBOL_LEVEL_UNCHANGED = -1
+} SYMBOL_LEVEL;
```
Comment (seanbudd) [inline nvdaHelper/interfaces/nvdaController/nvdaController.idl:54]: could you add a comment on this Enum that it must match the enum in python code.
Same with in the python definition, so that these 2 enums are kept in sync.

File: user_docs/en/changes.t2t [hunk 6]
```diff
@@ -70,7 +70,18 @@ Setting ``numCells`` is still supported for single line braille displays and ``n
  - Example: ``speech.speakSsml('<speak><prosody pitch="200%">hello</prosody><break time="500ms" /><prosody rate="50%">John</prosody></speak>')``
  - The SSML parsing capabilities are backed by the ``SsmlParser`` class in the ``speechXml`` module.
  -
-- The file names of the NVDA Controller Library no longer contain a suffix denoting the architecture, i.e. ``nvdaControllerClient32/64.dll`` are now called ``nvdaControllerClient.dll``. (#15718, #15717, @LeonarddeR)
+- Changes to the NVDA Controller Client library:
+ - The file names of the Library no longer contain a suffix denoting the architecture, i.e. ``nvdaControllerClient32/64.dll`` are now called ``nvdaControllerClient.dll``. (#15718, #15717, @LeonarddeR)
```
Comment (seanbudd) [inline user_docs/en/changes.t2t:74]:
```suggestion
 - The file names of the library no longer contain a suffix denoting the architecture, i.e. ``nvdaControllerClient32/64.dll`` are now called ``nvdaControllerClient.dll``. (#15718, #15717, @LeonarddeR)
```

File: user_docs/en/changes.t2t [hunk 7]
```diff
@@ -70,7 +70,18 @@ Setting ``numCells`` is still supported for single line braille displays and ``n
...
+- Changes to the NVDA Controller Client library:
+ - The file names of the Library no longer contain a suffix denoting the architecture, i.e. ``nvdaControllerClient32/64.dll`` are now called ``nvdaControllerClient.dll``. (#15718, #15717, @LeonarddeR)
+ - Added the following functions to the controller client: (#15734, #11028, #5638, @LeonarddeR)
+  - ``nvdaController_getProcessId``: To get the process id (PID) of the current instance of NVDA the controller client is using.
+  - nvdaController_speakSsml: To instruct NVDA to speak according to the given SSML. This function also supports:
+   - Providing the symbol level.
+   - Providing the priority of speech to be spoken.
+   - speaking both synchronously (blockking) and asynchronously (instant return).
+   - The exported ``_nvdaController_onSsmlMarkReached`` pointer can be assigned to a callback function that is called in synchronous mode for every ``<mark />`` tag encountered in the SSML sequence.
+   -
```
Comment (seanbudd) [inline user_docs/en/changes.t2t:82]: Fix indentation, added list terminator, fix typos
```suggestion
    - Providing the symbol level.
    - Providing the priority of speech to be spoken.
    - Speaking both synchronously (blocking) and asynchronously (instant return).
    - The exported ``_nvdaController_onSsmlMarkReached`` pointer can be assigned to a callback function that is called in synchronous mode for every ``<mark />`` tag encountered in the SSML sequence.
    -
  -
```

File: user_docs/en/changes.t2t [hunk 8]
```diff
@@ -70,7 +70,18 @@ Setting ``numCells`` is still supported for single line braille displays and ``n
  - Example: ``speech.speakSsml('<speak><prosody pitch="200%">hello</prosody><break time="500ms" /><prosody rate="50%">John</prosody></speak>')``
  - The SSML parsing capabilities are backed by the ``SsmlParser`` class in the ``speechXml`` module.
  -
-- The file names of the NVDA Controller Library no longer contain a suffix denoting the architecture, i.e. ``nvdaControllerClient32/64.dll`` are now called ``nvdaControllerClient.dll``. (#15718, #15717, @LeonarddeR)
+- Changes to the NVDA Controller Client library:
+ - The file names of the Library no longer contain a suffix denoting the architecture, i.e. ``nvdaControllerClient32/64.dll`` are now called ``nvdaControllerClient.dll``. (#15718, #15717, @LeonarddeR)
+ - Added the following functions to the controller client: (#15734, #11028, #5638, @LeonarddeR)
+  - ``nvdaController_getProcessId``: To get the process id (PID) of the current instance of NVDA the controller client is using.
+  - nvdaController_speakSsml: To instruct NVDA to speak according to the given SSML. This function also supports:
```
Comment (seanbudd) [inline user_docs/en/changes.t2t:77]:
```suggestion
  - ``nvdaController_speakSsml``: To instruct NVDA to speak according to the given SSML. This function also supports:
```

File: user_docs/en/changes.t2t [hunk 9]
```diff
@@ -74,7 +74,19 @@ Setting ``numCells`` is still supported for single line braille displays and ``n
...
-- The file names of the NVDA Controller Library no longer contain a suffix denoting the architecture, i.e. ``nvdaControllerClient32/64.dll`` are now called ``nvdaControllerClient.dll``. (#15718, #15717, @LeonarddeR)
+- Changes to the NVDA Controller Client library:
+ - The file names of the library no longer contain a suffix denoting the architecture, i.e. ``nvdaControllerClient32/64.dll`` are now called ``nvdaControllerClient.dll``. (#15718, #15717, @LeonarddeR)
+ - Added the following functions to the controller client: (#15734, #11028, #5638, @LeonarddeR)
+  - ``nvdaController_getProcessId``: To get the process id (PID) of the current instance of NVDA the controller client is using.
+  - ``nvdaController_speakSsml``: To instruct NVDA to speak according to the given SSML. This function also supports:
+    - Providing the symbol level.
+    - Providing the priority of speech to be spoken.
+    - Speaking both synchronously (blocking) and asynchronously (instant return).
+    - The exported ``_nvdaController_onSsmlMarkReached`` pointer can be assigned to a callback function that is called in synchronous mode for every ``<mark />`` tag encountered in the SSML sequence.
```
Comment (lukaszgo1) [inline user_docs/en/changes.t2t:85]: It is very likely my lack of understanding of how RPC rather than a real problem, but looking at this I cannot figure out how two applications wanting to provide their own callback for `onSsmlMarkReached` can do so. Since it seems to be necessary to assign the callback to the global pointer, it appears that, to be on the safe side, application developer should assign the old  value of `onSsmlMarkReached` to a temporary variable before passing any speech to NVDA, assign their own callback, speak whatever has to be spoken, and restore the old value afterwards. Would it be possible to pass pointer to `onSsmlMarkReached` as an additional parameter to `nvdaController_speakSsml` instead of using the single global callback?

**Comment by lukaszgo1** [comment]
> Note that this doesn't work very well with Espeak, but that's more of espeaks fault. This is best tested with OneCore.


Is this caused by issues in eSpeak, or our driver for it. In other words should this be reported against NVDA (if so is there an existing issue), or eSpeak (same question)?

> Version 1.0 supports all past and present versions of NVDA, including 2024.1.


Would it be possible to create a system test bundling version 1.0 of controller client, which ensures that the current NVDA can indeed speak through it?

> Version 1.1 is linked in the readme.

I assume you mend version 1.0 here?

File: nvdaHelper/interfaces/nvdaController/nvdaController.idl (hunk 4 above)
Comment (seanbudd) [reply nvdaHelper/interfaces/nvdaController/nvdaController.idl]: It seems like these comments are missing from the python module

File: nvdaHelper/interfaces/nvdaController/nvdaController.idl (hunk 5 above)
Comment (seanbudd) [reply nvdaHelper/interfaces/nvdaController/nvdaController.idl:62]: It seems like these comments are missing from the python module

File: source/NVDAHelper.py [hunk 10]
```diff
@@ -78,50 +79,68 @@ def nvdaController_speakText(text):
...
 @WINFUNCTYPE(c_long, c_wchar_p, c_int, c_int, c_bool)
-def nvdaController_speakSsml(
+def nvdaController_speakSsml(  # noqa: C901
 		ssml: str,
-		symbolLevel: int,
+		symbolLevel: "SymbolLevel",
 		priority: "SpeechPriority",
 		asynchronous: bool,
-):
+) -> int | SystemErrorCodes:
```
Comment (seanbudd) [inline source/NVDAHelper.py:91]: I think since the `-1` value was removed this is the return type now?

```suggestion
) -> SystemErrorCodes:
```

File: source/NVDAHelper.py (hunk 3 above)
Comment (seanbudd) [reply source/NVDAHelper.py:174]: I think this is still an issue

**Comment by leonardder** [comment]
> > Note that this doesn't work very well with Espeak, but that's more of espeaks fault. This is best tested with OneCore.
> 
> Is this caused by issues in eSpeak, or our driver for it. In other words should this be reported against NVDA (if so is there an existing issue), or eSpeak (same question)?

I recall that @jcsteh bumped into issues with Espeak indexing during speech refactor, but I"m not sure. I'm pretty sure it is an issue in ESpeak though.

> > Version 1.0 supports all past and present versions of NVDA, including 2024.1.
> 
> Would it be possible to create a system test bundling version 1.0 of controller client, which ensures that the current NVDA can indeed speak through it?

I'm not very comfortable with writing system tests honestly. I guess that this can be a unit test though.

> > Version 1.1 is linked in the readme.
> 
> I assume you mend version 1.0 here?

Correct, sorry.

File: nvdaHelper/interfaces/nvdaController/nvdaController.idl (hunk 5 above)
Comment (leonardder) [reply nvdaHelper/interfaces/nvdaController/nvdaController.idl:62]: Ah I'm sorry, I missed your second question there.

File: user_docs/en/changes.t2t (hunk 9 above)
Comment (leonardder) [reply user_docs/en/changes.t2t:91]: > it appears that, to be on the safe side, application developer should assign the old value of `onSsmlMarkReached` to a temporary variable before passing any speech to NVDA, assign their own callback, speak whatever has to be spoken, and restore the old value afterwards. 

That's not necessary. The old value is nullptr anyway and the internal nvdaController_onSsmlMarkReached only calls the exported pointer when it is set correctly.

> Would it be possible to pass pointer to `onSsmlMarkReached` as an additional parameter to `nvdaController_speakSsml` instead of using the single global callback?

That would be a lot nicer indeed, though I don't think we'd be able to pass the function pointer over RPC.

I'll have a look into this.

Comment (lukaszgo1) [reply user_docs/en/changes.t2t:91]: > > it appears that, to be on the safe side, application developer should assign the old value of `onSsmlMarkReached` to a temporary variable before passing any speech to NVDA, assign their own callback, speak whatever has to be spoken, and restore the old value afterwards.
> 
> That's not necessary. The old value is nullptr anyway and the internal nvdaController_onSsmlMarkReached only calls the exported pointer when it is set correctly.


Backing up and restoring the old function pointer seems to still be necessary, without doing so I see no way in which two applications can provide their own callback. Without backing up and restoring, the last assigned callback will always be used.

**Comment by lukaszgo1** [comment]
> > Would it be possible to create a system test bundling version 1.0 of controller client, which ensures that the current NVDA can indeed speak through it?
> 
> I'm not very comfortable with writing system tests honestly. I guess that this can be a unit test though.


Doing this as an unit test seems tricky as you need to have `NVDAHelper` and speech initialized. In system tests the entire screen reader is running, so it appears to be much easier.

File: user_docs/en/changes.t2t (hunk 9 above)
Comment (leonardder) [reply user_docs/en/changes.t2t:91]: Every application can provide its own callback because the callback is stored in the user space of the client. The server, NVDA in this case, doesn't know anything about what the callback does. A callback is just RPC in the opposite direction.

**Comment by leonardder** [comment]
I found out a nice way to avoid the API breakage. I will update the description, changelog and docs accoordingly later this week.

File: nvdaHelper/local/rpcSrv.cpp [hunk 11]
```diff
@@ -27,7 +27,8 @@ using namespace std;
 typedef RPC_STATUS(RPC_ENTRY *RpcServerRegisterIf3_functype)(RPC_IF_HANDLE,UUID __RPC_FAR*,RPC_MGR_EPV __RPC_FAR*,unsigned int,unsigned int,unsigned int,RPC_IF_CALLBACK_FN __RPC_FAR*,void __RPC_FAR*);
 
 RPC_IF_HANDLE availableInterfaces[]={
-	nvdaController_NvdaController_v1_1_s_ifspec,
+	nvdaController_NvdaController_v1_0_s_ifspec,
+	nvdaController_NvdaController2_v1_0_s_ifspec,
```
Comment (seanbudd) [inline nvdaHelper/local/rpcSrv.cpp:31]: shouldn't this be v2.0 if thats how it is being described in the readme?

```suggestion
	nvdaController_NvdaController_v2_0_s_ifspec,
```

File: nvdaHelper/interfaces/nvdaController/nvdaController.idl [hunk 12]
```diff
@@ -62,11 +62,25 @@ typedef [v1_enum] enum {
...
 /**
- * Allows the controling of NVDA from a remote process
+ * Type signature for a callback that may be implemented by a client.
+ * The callback is called for every mark in the SSML as received by speakSsml when called synchronously.
+ * @param mark The name of the reached mark.
+ */
+typedef error_status_t(__stdcall *onSsmlMarkReachedFuncType)([in, string] const wchar_t* mark);
+
+/**
+ * Setts the callback used when marks are reached within SSML.
```
Comment (seanbudd) [inline nvdaHelper/interfaces/nvdaController/nvdaController.idl:72]:
```suggestion
 * Sets the callback used when marks are reached within SSML.
```

File: nvdaHelper/interfaces/nvdaController/nvdaController.idl [hunk 13]
```diff
@@ -62,11 +62,25 @@ typedef [v1_enum] enum {
...
+ * The callback is called for every mark in the SSML as received by speakSsml when called synchronously.
+ * Param callback The callback to use, NULL to reset.
```
Comment (seanbudd) [inline nvdaHelper/interfaces/nvdaController/nvdaController.idl:74]:
```suggestion
 * @param callback: The callback to use, NULL to reset.
```

File: nvdaHelper/interfaces/nvdaController/nvdaController.idl [hunk 14]
```diff
@@ -62,11 +62,25 @@ typedef [v1_enum] enum {
 } SYMBOL_LEVEL;
 
 /**
- * Allows the controling of NVDA from a remote process
+ * Type signature for a callback that may be implemented by a client.
+ * The callback is called for every mark in the SSML as received by speakSsml when called synchronously.
+ * @param mark The name of the reached mark.
```
Comment (seanbudd) [inline nvdaHelper/interfaces/nvdaController/nvdaController.idl:67]: I find it hard to parse these without a separator value between the parameter name and description. I'm not sure what the standard is, but a colon makes it much more readable in my opinion

```suggestion
 * @param mark: The name of the reached mark.
```

File: nvdaHelper/interfaces/nvdaController/nvdaController.idl [hunk 15]
```diff
@@ -91,9 +105,19 @@ interface NvdaController {
...
+ * Adds additional methods to control NVDA from a remote process
+ */
+[
+	uuid(3D168D45-CB58-4270-8257-4E0BE515D557),
+	version(1.0),
+]
+interface NvdaController2 {
 /**
  * Retrieves the process identifier (PID) of NVDA's process.
+ * @param pid Out parameter that receives the process id.
```
Comment (seanbudd) [inline nvdaHelper/interfaces/nvdaController/nvdaController.idl:120]:
```suggestion
 * @param pid: Out parameter that receives the process id.
```

File: extras/controllerClient/readme.md [hunk 16]
```diff
@@ -41,14 +41,17 @@ You can build locally or download pre-built, details:
 * `nvdaController.h`
   - A C header file containing declarations for all the provided functions.
 
-The **`extras/controllerClient/x86` directory** also contains example usage:
+The **`extras/controllerClient/Examples` directory** also contains example usage:
```
Comment (seanbudd) [inline extras/controllerClient/readme.md:44]: can the Examples folder be renamed to lowerCamelCase? This is more standard in the folder structure.

```suggestion
The **`extras/controllerClient/examples` directory** also contains example usage:
```

File: extras/controllerClient/Examples/example_csharp/SymbolLevel.cs [hunk 17]
```diff
@@ -0,0 +1,16 @@
+﻿// Copyright(C) 2017-2023 NV Access Limited, Arnold Loubriat, Leonard de Ruijter
+// This file is covered by the GNU Lesser General Public License, version 2.1.
+// See the file license.txt for more details.
+
+namespace NVAccess.NVDA
+{
+    public enum SymbolLevel
```
Comment (seanbudd) [inline extras/controllerClient/Examples/example_csharp/SymbolLevel.cs:7]: it may be also worth referencing the python enum here, even though this is an example. I don't think a 2-way reference is needed, as it can be the reader's responsibility to check the other enums and ensure this example reflects them correctly.

File: extras/controllerClient/Examples/example_csharp/SpeechPriority.cs [hunk 18]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright(C) 2017-2023 NV Access Limited, Arnold Loubriat, Leonard de Ruijter
+// This file is covered by the GNU Lesser General Public License, version 2.1.
+// See the file license.txt for more details.
+
+namespace NVAccess.NVDA
+{
+    public enum SpeechPriority
```
Comment (seanbudd) [inline extras/controllerClient/Examples/example_csharp/SpeechPriority.cs:7]: it may be also worth referencing the python enum here, even though this is an example. I don't think a 2-way reference is needed, as it can be the reader's responsibility to check the other enums and ensure this example reflects them correctly.

File: extras/controllerClient/Examples/example_csharp/Program.cs.cs [hunk 19]
```diff
@@ -0,0 +1,61 @@
+﻿// Copyright(C) 2017-2023 NV Access Limited, Arnold Loubriat, Leonard de Ruijter
```
Comment (seanbudd) [inline extras/controllerClient/Examples/example_csharp/Program.cs.cs:1]: this file is named with 2 file extensions, `.cs.cs` is that intentional?

File: extras/controllerClient/Examples/example_csharp/NVDA.cs [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+        }
+
+        /// <summary>
+        /// Instructs NVDA to speak the given Speech Synthesis Markup Language (SSML).
+        /// </summary>
+        /// <param name="ssml">The ssml to speak.</param>
+        /// <param name="symbolLevel">The symbol verbosity level.</param>
+        /// <param name="priority">The priority of the speech sequence.</param>
+        /// <param name="asynchronous">Whether SSML should be spoken asynchronously.</param>
+        public static void SpeakSsml(string ssml, SymbolLevel symbolLevel = SymbolLevel.Unchanged, SpeechPriority priority = SpeechPriority.Normal, bool asynchronous = true, OnSsmlMarkReached callback = null)
```
Comment (seanbudd) [inline extras/controllerClient/Examples/example_csharp/NVDA.cs:97]: could this be linted onto new lines please

File: nvdaHelper/local/rpcSrv.cpp (hunk 11 above)
Comment (seanbudd) [reply nvdaHelper/local/rpcSrv.cpp:31]: Thoughts on this issue?

File: nvdaHelper/interfaces/nvdaController/nvdaController.idl (hunk 14 above)
Comment (seanbudd) [reply nvdaHelper/interfaces/nvdaController/nvdaController.idl:67]: Thoughts on this suggestion?

**Review by LeonarddeR — COMMENTED** [review]
I'm sorry, I had pending comments I forgot to post yesterday.

**Review by seanbudd — APPROVED** [review]

**Comment by leonardder** [comment]
Thanks for approving @seanbudd.  Is this still blocked by something?

File: nvdaHelper/interfaces/nvdaController/nvdaController.idl (hunk 14 above)
Comment (michaelDCurran) [reply nvdaHelper/interfaces/nvdaController/nvdaController.idl:67]: My preference would be to keep using the original doc formatting (I.e. with out the colon) so that it remains standard across the codebase. However, I would nto be against changing the syntax  of all c++ code docs to something more readable, as long as tools like doxygen or an equivalent could parse it, and that we do it all in one go.

File: nvdaHelper/local/rpcSrv.cpp (hunk 11 above)
Comment (michaelDCurran) [reply nvdaHelper/local/rpcSrv.cpp:31]: Agreed with @leonardder here. The specific interface version is somewhat internal and should be 1.0 as it is the first version. the public client can be called version 2, that is fine with me.

File: source/NVDAHelper.py [hunk 21]
```diff
@@ -63,31 +74,118 @@ def nvdaController_speakText(text):
...
+
+	try:
+		priority = SpeechPriority(priority)
+	except ValueError:
+		log.exception("Invalid SpeechPriority")
+		return SystemErrorCodes.INVALID_PARAMETER
+
+	prefixSpeechCommand = None
+	markCallable = None
+	if not asynchronous:
```
Comment (michaelDCurran) [inline source/NVDAHelper.py:115]: Am I write in understanding that if the function is called with asynchronous as True, the mark callback will not be ever fired?

Comment (leonardder) [reply source/NVDAHelper.py:115]: This is correct. As per the [docs about callbacks](https://learn.microsoft.com/en-us/windows/win32/midl/callback):

> The callback function ... lets the server query the client for information in the context of the original call. Callbacks are special cases of remote calls that execute as part of a single thread. A callback is issued in the context of a remote call. Any remote procedure defined as part of the same interface as the static callback function can call the callback function.

So in other words, a callback can only be called during a call. Calling the callback as part of an async call that was already completed wouldn't work.
